src/Authenticator/EnvironmentAuthenticator.php

```diff
@@ -0,0 +1,127 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         2.10.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Authentication\Authenticator;
+
+use Authentication\UrlChecker\UrlCheckerTrait;
+use Psr\Http\Message\ServerRequestInterface;
+
+/**
+ * Form Authenticator
+ *
+ * Authenticates an identity based on the POST data of the request.
+ */
+class EnvironmentAuthenticator extends AbstractAuthenticator
+{
+    use UrlCheckerTrait;
+
+    /**
+     * Default config for this object.
+     * - `fields` The fields to use to identify a user by.
+     * - `loginUrl` Login URL or an array of URLs.
+     * - `urlChecker` Url checker config.
+     *
+     * @var array
+     */
+    protected $_defaultConfig = [
+        'loginUrl' => null,
+        'urlChecker' => 'Authentication.Default',
+        'attributes' => [],
```

Comment:
Why is this `attributes` and not `fields`?

Reply:
I've hesitated about the term. I chose `attributes` to stay close to the `Shibboleth` documentation. When you setup `Shibboleth` you choose what attributes you want, but you're right, it might be confusing.

Reply:
Now, you made me think that what I did is not necessary. The `attributes` parameter (renamed as `fields`)  should contain a list of environment variable names one would like to grab, without mapping names as well. Better have identifiers map values no?

Reply:
I'm not well versed with Shibboleth. But we'll you need a way to map the environment variables to the fields in your database schema that identify users.

For other parts of authentication we use `fields` to connect the form fields/request data to the username/password/token credential fields. Would something like 

```php
'fields' => [IdentifierInterface::CREDENTIAL_TOKEN => 'SHIBBOLETH_ID']
```
work? My question about `attributes` was trying to see if we need more than just the standard identifier/authenticator fields.

Reply:
Indeed it could work, but I have a use case where it wouldn't, where a person is identified the first time by their email, and then by their unique_id .

That application manages our students. Only known students can access it. There is no registration form. Students are added to the database by a supervisor, and are identified at first by their private email, as it is not possible to know their SHIBBOLETH_ID before they connect for the first time.
Multiple emails can be linked to a single SHIBBOLETH_ID, that's why I grab and save it after their first identification. Subsequent identification are done against it and not against their email anymore 

Thinking of it, could I load the `EnvironmentAuthenticator` twice, with each a different `CREDENTIAL_TOKEN`? I'll try it, and if that works, I'll answer yes to your question. I'll try that tomorrow.

The identity provider (Shibboleth) can also provide different attributes/fields that are related to the identified person that can be necessary to have for an other purpose then identification. In my case, among others, I can detect when a student is registered, or when he left the program.